optimum/habana/transformers/trainer.py

```diff
@@ -116,52 +124,58 @@
 if is_datasets_available():
     import datasets
 
-
 if is_safetensors_available():
     import safetensors.torch
 
-
 if is_peft_available():
     from peft import PeftModel
     from peft.utils import PeftType
 
-
 if is_deepspeed_available():
     from accelerate.utils import DeepSpeedSchedulerWrapper
 
-if is_accelerate_available():
-    from accelerate.utils import (
-        load_fsdp_optimizer,
-        save_fsdp_optimizer,
-    )
-
-if TYPE_CHECKING:
-    import optuna
+if is_accelerate_available("0.28.0"):
+    from accelerate.utils import DataLoaderConfiguration
 
 
-DATA_SAMPLERS = [RandomSampler, SeedableRandomSampler]
+def _get_input_update_settings(model) -> Tuple[bool, Dict]:
+    """
+    Determines whether the input settings need to be updated.
 
+    Currently (attn_softmax_bf16, use_flash_attention, flash_attention_recompute,
+    flash_attention_causal_mask) are enabled only for llama, qwen2, starcoder2 and gemma
 
-if is_accelerate_available("0.28.0"):
-    from accelerate.utils import DataLoaderConfiguration
+    Args:
+        model: The model instance for which the input update settings are being evaluated
 
+    Returns:
+        Tuple[bool, Dict]: A flag indicating whether the input settings should be updated.
+        A dictionary containing the specific input settings that need to be updated, if any
+    """
+    should_update_inputs: bool = (getattr(model, "generation_config", None) is not None) and (
+        model.config.model_type in ("llama", "qwen2", "starcoder2", "gemma")
+    )
 
-def _is_peft_model(model):
-    if is_peft_available():
-        classes_to_check = (PeftModel,) if is_peft_available() else ()
-        # Here we also check if the model is an instance of `PeftMixedModel` introduced in peft>=0.7.0: https://github.com/huggingface/transformers/pull/28321
-        if version.parse(importlib.metadata.version("peft")) >= version.parse("0.7.0"):
-            from peft import PeftMixedModel
+    inputs_update: Dict = {}
+    if should_update_inputs:
+        if model.generation_config.attn_softmax_bf16:
+            inputs_update["attn_softmax_bf16"] = True
+        if model.generation_config.use_flash_attention:
+            inputs_update["use_flash_attention"] = True
+        if model.generation_config.flash_attention_recompute:
+            inputs_update["flash_attention_recompute"] = True
+        if model.generation_config.flash_attention_causal_mask:
+            inputs_update["flash_attention_causal_mask"] = True
+        should_update_inputs = len(inputs_update) > 0
 
-            classes_to_check = (*classes_to_check, PeftMixedModel)
-        return isinstance(model, classes_to_check)
-    return False
+    return should_update_inputs, inputs_update
 
 
 if TYPE_CHECKING:
```

Comment:
It looks like the transformers code still has this code under TYPE_CHECKING.

if TYPE_CHECKING:
    import optuna

    if is_datasets_available():
        import datasets

Reply:
Yes, you are right. I have removed datasets in TYPE_CHECKING, since for transfomers it is optional but for optimum-habana it is a requirement, is imported above, thus there is no need to do it again.